{config,server}: Handle parsing, storing, and building embedded pipeline graphs

briancain opened the pull request:
This pull request introduces support to parse an embedded pipeline, where that pipeline could either be a Pipeline reference by Owner, or a nested in-line defined pipeline within the step stanza. Currently this PR supports a single embedded pipeline for parsing HCL, however the HCL to Proto func supports infinitely nested pipelines in the case that we might support that within an HCL config. For now, we'll artificially allow for a single nested pipeline until we can get a better user story for Many embedded pipelines within a single Step.

When a runner or user goes to sync pipeline configs, the PipelineProto func will now extract any nested/embedded defined Pipelines and store them as separate proto messages. The step where that embedded pipeline was defined then turns into a Pipeline reference step, which can be looked up at execution time.

This PR updates RunPipeline to build an entire pipeline graph including any nested/embedded pipeline graphs and its dependencies. It handles embedded pipeline reference cycles when queueing a pipeline from RunPipeline.

I've been using this project to test this pull request feature: https://github.com/briancain/waypoint-tetris

briancain committed:
internal/config: Add support for single nested pipeline in a step

This commit adds support for nesting an embedded pipeline within a step
of a parent pipeline.

briancain committed:
server/proto: Add a pipeline reference to Steps

briancain committed:
server: proto gen

briancain committed:
internal/config: PipelineProtos translates nested pipelines

This commit updates the config sync func PipelineProtos to read a nested
pipeline, either defined directly within a Step stanza, or as a simple
pipeline Owner ref, and turns that HCL data into Pipeline protobufs. If
an externally referenced pipeline is defined, the step will be a
Pipeline step ref to be looked up later at execution time. If the
pipeline is defined directly inside the step stanza, this function will
configure the Pipeline separately as its own pipeline, and update the
step stanza it was defined in to reference that pipeline.

This commit does not check for cycles yet.

briancain committed:
internal/server: Err on RunPipeline if nested Pipeline is set

This commit returns an Unimplemented error if a user happens to try to
queue a pipeline run that references an embedded pipeline. This will be
supported in a future commit.

briancain committed:
internal/config: Improve code doc comments

briancain committed:
internal/config: Invalid step returns an error

briancain committed:
server: Add test for PipelineGet to validate cycle detection

briancain committed:
server: Extract building pipeline proto to queue job request func

briancain committed:
server: Implement part of generating embedded pipeline jobs for
RunPipeline

This commit mostly implements the RunPipeline code for queueing a
pipeline that also embeds a pipeline. There are still several todos, so
this is mostly a partial commit.

briancain commented:
I decided to continue with part 2 on the same branch for now, I tried to leave TODOs in places where I'm still working. The config parsing and proto generation is ready for review though.

briancain committed:
server: Build dependency graph across pipelines prior to Run

This commit builds the full Pipeline run graph with associated jobs
prior to running the pipeline. It will also detect cycles between
embedded pipelines if any exist. To prevent cycles while building a
dependency graph, we include visisted nodes and pipelines so that if we
visit a node we've already been to, we simply move on rather than
continuing through the cycle

briancain commented:
FYI: My `serivce_pipeline` cycle test seems to randomly fail locally so I'm going to investigate that.

briancain committed:
server: Handle building graphs with same node names

Prior to this commit, building a graph of pipelines and embedded
pipelines had an issue if the nested pipeline steps had step names that
matched any other step from another pipeline. This in practice is valid
to have step names be similar to other pipelines but when we combine
them and use them to constrct a graph to determine dependencies, we
can't re-use that name because we expect node names in the graph to be
unique.

This commit fixes this by assigning node IDs to each node in a graph,
and storing a pipeline step ref to node Id map that we can look up later
when we need the step and pipeline names.

briancain committed:
server: Don't include embedded pipeline step refs as a job id

We include references to embedded pipeline steps for building a graph,
but it doesn't actually count as a "job" so don't include it in the job
map or all job ids slice returned from running a pipeline.

briancain committed:
config: Improve validation errors for various use stanza mistakes

briancain committed:
config: Validate when use and pipeline stanzas are present in stepraw

briancain committed:
config/validate: Only validate step uses if no pipeline defined

briancain committed:
lint: Fix spacing in long err message funcs

briancain committed:
server: Return cycle error if we've visisted a step in a pipeline

Also add more doc comments

briancain committed:
server/test: Add a deeply cyclic graph to ensure no cycles allowed

briancain committed:
internal/config: Change how to reference external pipeline in embed

This commit updates the syntax for referencing a pipeline by introducing
a built-in step named 'pipeline' with params 'project' and pipeline
'name'. This syntax is clearer for those compared to the magic
'project/pipeline-name' that it was before.

paladin-devops commented on internal/config/pipeline.go:
```diff
@@ -99,8 +104,48 @@ func (c *Config) Pipeline(id string, ctx *hcl.EvalContext) (*Pipeline, error) {
+			// Parse all the steps
+			var embSteps []*Step
+			for _, embedStepRaw := range embedPipeline.StepRaw {
+				// TODO(briancain): Do we add an error here if we detect another level
```
In case you're looking for an answer for this PR, then I think yes we should, so long as embedded pipelines won't work right now - which I think is the case until this function is recursive (as you mentioned in your earlier comment).

briancain replied:
Done!

paladin-devops commented on internal/config/pipeline.go:
```diff
@@ -118,148 +166,244 @@ func (c *Config) PipelineProtos() ([]*pb.Pipeline, error) {
+		// defined else where. If this is a ref, the raw hcl for the pipeline should
+		// be empty.
+		if step.PipelineRaw != nil {
+			// TODO(briancain): It shouldn't be valid to have a "project/pipeline"
```
Does this mean that a nested pipeline should not be able to refer to a pipeline in another project?

briancain replied:
Ah, so this comment is no longer valid! This was still around when I had external pipeline references as

```hcl
step "external-pipeline" {
  pipeline "project/pipeline" { }
  // This shouldn't be valid to have this too
  pipeline "test" {
    // config
  }
}
```

But got some feedback that the pipeline ref was too magical, so now you reference a pipeline with a step:

```hcl
step "external-pipeline-ref" {
  use "pipeline" {
    project = "test"
    name    = "my-pipeline"  
  }
}
```

So yeah, I'll remove this comment!

paladin-devops commented on internal/config/validate.go:
```diff
@@ -327,9 +327,15 @@ func (c *Pipeline) Validate() error {
-		if stepRaw == nil || stepRaw.Use == nil || stepRaw.Use.Type == "" {
+		if stepRaw == nil && stepRaw.PipelineRaw == nil {
 			result = multierror.Append(result, fmt.Errorf(
 				"step stage with a default 'use' stanza is required"))
```
```suggestion
				"step stage with a default 'use' stanza or a 'pipeline' stanza is required"))
```

paladin-devops commented on internal/config/validate_test.go:
```diff
@@ -65,6 +65,10 @@ func TestConfigValidatePipelines(t *testing.T) {
 			"",
 		},
+		{
+			"pipeline_nested_pipes.hcl",
```
Should the test files `pipeline_nested.hcl` and `pipeline_nested_refs.hcl` also be included?

briancain committed:
config/validate: Add nested pipeline ref hcl for validate test

briancain committed:
config/pipeline: Remove stale comment about pipeline ref

briancain committed:
Update internal/config/validate.go

Co-authored-by: Joe <83741749+paladin-devops@users.noreply.github.com>

briancain committed:
config/pipeline: Raise error if 2 levels of pipe nesting found

briancain committed:
server: Ensure we don't add pipeline refs as job templates

This commit ensures that we don't attempt to create job templates for a
pipeline ref step as it's not an operation but a reference to another
pipeline.

briancain committed:
server: Ensure embedded pipeline job templates set parent dependencies

This commit updates the pipeline run job template generator to include
any parent step job ID dependencies when generating childen jobs.
Otherwise prior to this, these jobs were scheduled with 0 dependencies
and would run prior to any previous dependent steps.

briancain commented:
Rebased off main to fix merge conflicts

briancain committed:
fix: Fix casing in err msg

briancain committed:
fix: Fix code comment about pipeline ref step

briancain committed:
fix: Clarify comments about parsing steps and embedded pipelines

briancain commented:
Rebased off main which includes pipeline run tracking

briancain committed:
maint: Ensure pipeline run sequence is set on step ref

paladin-devops commented on internal/config/pipeline.go:
```diff
@@ -99,8 +104,68 @@ func (c *Config) Pipeline(id string, ctx *hcl.EvalContext) (*Pipeline, error) {
+			var embSteps []*Step
+			for _, embedStepRaw := range embedPipeline.StepRaw {
+				if embedStepRaw.PipelineRaw != nil {
+					// NOTE(briancain): For now, we artificially don't allow for more than
```
Thank you for explicitly writing this comment!

paladin-devops replied:
And the status message! ✨

paladin-devops commented on internal/server/boltdbstate/job.go:
```diff
@@ -1321,6 +1321,11 @@ func (s *State) jobCreate(dbTxn *bolt.Tx, memTxn *memdb.Txn, jobpb *pb.Job) erro
 		for _, id := range jobpb.DependsOn {
+			if id == "" {
+				return status.Errorf(codes.FailedPrecondition,
+					"job %q has an empty string id for depends_on", jobpb.Id)
```
Good catch! I'm wondering how this could happen... I can't recall when I've seen a job without an ID 🤔

briancain replied:
This came about from a mistake in my own code! I was adding empty string to the `DependsOn` slice in the job template generator (this was a bug) and figured we could use some extra validation here that this shouldn't be allowed.

izaaklauer reviewed (approved):
I haven't tested this, but the approach and the code look pretty good to me! Also looks well-tested 👍

briancain committed:
cli/init: Auto-sync pipeline protos from config

This commit automatically syncs all defined pipeline Protobuf values
when running `waypoint init` for a new project similar to how it syncs
projects and applications automatically

briancain committed:
cli/pipeline_run: Run ConfigSync prior to running a pipeline

This commit ensures that the Waypoint server has a proper up to date
version of all pipelines in a project prior to them being run.

briancain committed:
website autogen

xiaolin-ninja commented on internal/config/validate.go:
```diff
@@ -327,9 +327,15 @@ func (c *Pipeline) Validate() error {
 
 	for _, stepRaw := range c.StepRaw {
-		if stepRaw == nil || stepRaw.Use == nil || stepRaw.Use.Type == "" {
+		if stepRaw == nil && stepRaw.PipelineRaw == nil {
```
The second part of this conditional would never get hit, since if `stepRaw == nil` then it would never evaluate `stepRaw.PipelineRaw`.
```suggestion
		if stepRaw == nil || stepRaw.PipelineRaw == nil {
```

briancain replied:
I think I'll want to split up the if-statement then because if `stepRaw` is nil, `stepRaw.PipelineRaw` will panic. But makes sense!

xiaolin-ninja commented on internal/config/validate.go:
```diff
@@ -327,9 +327,15 @@ func (c *Pipeline) Validate() error {
+				"step stage with a default 'use' stanza or a 'pipeline' stanza is required"))
+		} else if stepRaw.Use != nil && stepRaw.PipelineRaw != nil {
+			result = multierror.Append(result, fmt.Errorf(
+				"step stage with a 'use' stanza and pipeline stanza is not valid"))
```
This message could be clearer to express that you can't "use" a plugin _and_ have a nested "pipeline" stanza. (At least that's how I understood this code?)

briancain replied:
That's correct!

xiaolin-ninja commented on internal/config/validate.go:
```diff
@@ -327,9 +327,15 @@ func (c *Pipeline) Validate() error {
+				"step stage with a 'use' stanza and pipeline stanza is not valid"))
+		} else if stepRaw.PipelineRaw == nil && (stepRaw.Use == nil || stepRaw.Use.Type == "") {
+			result = multierror.Append(result, fmt.Errorf(
+				"step stage %q is required to define a 'use' stanza and label", stepRaw.Name))
```
Same here, I'm understanding the check as "if no nested pipeline or plugin is defined for the step", but the message doesn't express that.

xiaolin-ninja commented on pkg/server/proto/server.proto:
```diff
@@ -4705,6 +4705,9 @@ message Pipeline {
       // Built-in up operation
       Up up = 8;
 
+      // A reference to a different pipeline
```
```suggestion
      // A reference to an embedded pipeline
```
But also, why do we include the entire pipeline object here instead of just a ref? It would make GetPipeline endpoints return unnecessarily heavy content, no?

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
+	pipelineRun *pb.PipelineRun,
+) ([]*pb.QueueJobRequest, *pb.PipelineRun, map[string]string, error) {
+	if len(visitedPipelines) != 0 {
+		// Determine if we've already visisted this pipeline and included its jobs.
```
```suggestion
		// Determine if we've already visited this pipeline and included its jobs.
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+
+// pipelineGraphFull takes a pipeline, and optionally accepts an existing Graph
+// and parent step, and attempts to build a full graph for a given Pipeline including
+// any nested pipeline steps. It keeps track of visisted steps so that we don't
```
```suggestion
// any nested pipeline steps. It keeps track of visited steps so that we don't
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+	log hclog.Logger,
+	g *graph.Graph,
+	parentStep string,
+	visistedNodes map[string]string,
```
```suggestion
	visitedNodes map[string]string,
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+	// Note that pipeline.Steps is not an ordered list of steps. It's a map of key val
+	// steps so the order will not match the order steps are defined in a waypoint.hcl.
+	for _, step := range pipeline.Steps {
+		if len(visistedNodes) != 0 {
```
```suggestion
		if len(visitedNodes) != 0 {
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+			if pipeName, ok := visistedNodes[step.Name]; ok && pipeName == pipeline.Name {
+				log.Trace("we've cycled to a node we've already visisted!", "pipeline", pipeName, "step", step.Name)
+				return nil, nil, status.Error(codes.FailedPrecondition,
+					"we've already visisted this node, that means we've got a cycle")
```
```suggestion
					"we've already visited this node, that means we've got a cycle")
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+	// steps so the order will not match the order steps are defined in a waypoint.hcl.
+	for _, step := range pipeline.Steps {
+		if len(visistedNodes) != 0 {
+			if pipeName, ok := visistedNodes[step.Name]; ok && pipeName == pipeline.Name {
```
```suggestion
			if pipeName, ok := visitedNodes[step.Name]; ok && pipeName == pipeline.Name {
```
lol Brian. After seeing this word so many times even I'm like, "wait, how do you spell visited? Is visisted the actual word?"

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+	for _, step := range pipeline.Steps {
+		if len(visistedNodes) != 0 {
+			if pipeName, ok := visistedNodes[step.Name]; ok && pipeName == pipeline.Name {
+				log.Trace("we've cycled to a node we've already visisted!", "pipeline", pipeName, "step", step.Name)
```
```suggestion
				log.Trace("we've cycled to a node we've already visited!", "pipeline", pipeName, "step", step.Name)
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+
+		// Keep track of the fact that we've visited this step node in this pipeline
+		// to prevent infinite cycles as we build embedded pipeline graphs
+		visistedNodes[step.Name] = pipeline.Name
```
```suggestion
		visitedNodes[step.Name] = pipeline.Name
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+			// Build the nested pipelines graph
+			parentStep := nodeId
+			embeddedGraph, n, err := s.pipelineGraphFull(ctx, log, stepGraph,
+				parentStep, visistedNodes, nodeIdMap, embeddedPipeline)
```
```suggestion
				parentStep, visitedNodes, nodeIdMap, embeddedPipeline)
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
+		return nil, err
+	}
+
+	// Build out all of the queued job requests for running this pipelines steps
```
```suggestion
	// Build out all of the queued job requests for running this pipeline's steps
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -213,6 +352,36 @@ func (s *Service) RunPipeline(
+				return nil, nil, nil, err
+			}
+
+			// Pass through *this* steps job dependency IDs so that the child
+			// steps aren't scheduled prior to any dependencies they are waiting for.
```
```suggestion
			// Pass through *this* step's job dependency IDs, so that the child
			// step's jobs aren't scheduled prior to any dependencies.
```
Is this what you mean?

briancain replied:
Yep!

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -213,6 +352,36 @@ func (s *Service) RunPipeline(
+			// steps aren't scheduled prior to any dependencies they are waiting for.
+			parentStepDep := map[string][]string{pipeline.Id: job.DependsOn}
+
+			embedJobs, embedRun, embedStepIds, err := s.buildStepJobs(ctx, log, req,
```
A RECURSIVE FUNCTION?! 🤯

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
+	pipelineRun *pb.PipelineRun,
+) ([]*pb.QueueJobRequest, *pb.PipelineRun, map[string]string, error) {
+	if len(visitedPipelines) != 0 {
+		// Determine if we've already visisted this pipeline and included its jobs.
+		// Otherwise we'll get stuck in a cycle. This only really works because
+		// pipeline names are unique for a project. If we ever start allowing for
+		// pipelines across projects we'll need to namespace this value.
```
```suggestion
		// Determine if we've already visited this pipeline and included its jobs.
		// Otherwise, we'll get stuck in a cycle. This only works because
		// pipeline names are unique for a project. If we ever allow
		// pipelines across projects, we'll need to namespace this value.
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
+	}
+
+	// Mark that we've visited this pipeline already
+	visitedPipelines[pipeline.Name] = struct{}{}
```
I would use a boolean instead of an empty struct for more clarity and space efficiency, but it doesn't actually matter.

briancain replied:
This is actually an established pattern already with our graph package libraries:

https://github.com/hashicorp/waypoint/blob/4b93417dda83c54762d80a5d996e3916b2a8fe03/internal/pkg/graph/dijkstra.go#L40-L45

https://github.com/hashicorp/waypoint/blob/4b93417dda83c54762d80a5d996e3916b2a8fe03/internal/pkg/graph/dfs.go#L18-L20

I like that it's a nonsense type mostly because I could imagine a `bool` or other var type being a bit confusing for a future reader given that we don't care about the value in this map, we're using a map for quick lookups.

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
+			// NOTE(briancain):
+			// This could be better. It's basically here because we want to keep track
+			// of an embedded pipeline step ref within the step graph, but it doesn't have
+			// a "job id" because the root of the embedded pipeline is the actual ID where
```
```suggestion
			// a "job id" because the root of the embedded pipeline is the actual ID whereas
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
+	visitedPipelines[pipeline.Name] = struct{}{}
+
+	// Generate job IDs for each of the steps. We need to know the IDs in
+	// advance to setup the dependency chain.
```
```suggestion
	// advance to set up the dependency chain.
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
+			dependsOn = append(dependsOn, d)
+		}
+
+		// Depend on all JOB IDs from parent step. If we were given a parent pipeline
```
```suggestion
		// Depend on all job IDs from parent step. If we were given a parent pipeline
```

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
+	log hclog.Logger,
+	req *pb.RunPipelineRequest,
+	visitedPipelines map[string]interface{},
+	nodeIdMap map[string]*pb.Ref_PipelineStep,
```
Idea, what if we switched this to:

`map[*pb.Ref_PipelineStep]string`? Then instead of adding another nested loop in `stepToNodeId` to loop through the whole map for a value match, we can do an O(1) look up and grab the `nodeId`?

briancain replied:
I believe I initially attempted this structure, where a protobuf was the `key` for the map, but you can't safely compare protobuf messages as values I _don't think_. I know Golang does allow for regular Go structs to be used as keys for a map key, but as far as a Protobuf message goes it's recommended to use [proto.Equal](https://developers.google.com/protocol-buffers/docs/reference/go/faq#deepequal) for comparison and I'm not sure how that behaves with maps, thus the structure here.

```golang
stepRef := &pb.Ref_PipelineStep {
  Name: "test",
}
mymap := map[*pb.Ref_PipelineStep]string{stepRef: "123abc"}

// Not sure about this one:
fmt.Println(mymap[stepRef])
```

If this is safe and reliable to do though it would make this code simpler for sure and would solve the issue you mentioned later in this PR about the node ID lookup loop iterating over this map.

briancain replied:
Ok, @evanphx and I figured out a decent solution to this! Solved via [d524338](https://github.com/hashicorp/waypoint/pull/3584/commits/d5243388e6c1979bd0d49ca1ee71bea3b52368a7)

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -118,26 +117,163 @@ func (s *Service) RunPipeline(
+	stepIds := map[string]string{}
+	for name, step := range pipeline.Steps {
+		if _, ok := step.Kind.(*pb.Pipeline_Step_Pipeline_); !ok {
+			nodeId, ok := s.stepToNodeId(ctx, log, pipeline.Id, name, nodeIdMap)
```
This `s.stepToNodeId` look up is another loop, which makes this triple nested loop inside a recursive function, which feels _really_ heavy to me. What if we restructure NodeId map to use `Ref_PipelineStep` as the key, and the `nodeId` as value, which will save a lot of processing power? (Which could really matter, especially with HCP or a large nesting pipeline.)

briancain replied:
Solved via [d524338](https://github.com/hashicorp/waypoint/pull/3584/commits/d5243388e6c1979bd0d49ca1ee71bea3b52368a7)!

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+	if g != nil {
+		// We're handling an embedded pipeline graph
+		stepGraph = g
+	} else if stepGraph == nil {
+		stepGraph = &graph.Graph{}
 	}
```
What is the point of this `else` if `stepGraph` is declared as nil on line 416 and remains nil unless the conditional is hit?

briancain replied:
The main point here is initializing `stepGraph` from nil into an empty `*graph.Graph{}` struct in the base case or first iteration of this func that I'm handling. I figured it might be worth being explicit assigning `stepGraph` when it's nil rather than an empty struct, but might not matter?

xiaolin-ninja replied:
```suggestion
	stepGraph := &graph.Graph{}
	if g != nil {
		// We're handling an embedded pipeline graph
		stepGraph = g
```
Yea, the way the code is written now, it's superfluous. I would declare it as the empty graph if that's the desired state if `g != nil`.

briancain replied:
Yeah makes sense! I'll update it 👍🏻

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -221,45 +390,166 @@ func (s *Service) RunPipeline(
+func (s *Service) stepToNodeId(
+	ctx context.Context,
+	log hclog.Logger,
+	pipelineName string,
```
```suggestion
	pipelineId string,
```

`stepRef.Pipeline` is the ID of the pipeline, but the variable is `pipelineName`, but this method is always called with the input variable `pipeline.Id`. So we should rename this variable to be clearer.

_OR_ if we want to always reference the pipeline by Name and not ID, then we should refactor `Ref_PiplineStep`.

briancain replied:
Makes sense to me to rename this local var `pipelineId`!

xiaolin-ninja reviewed (changes requested):
Wow, this was mind-bending to read and clearly took a ton of work! Amazing job. 🥇 

I'm a bit concerned about the triple nested loops inside of a recursive function, and have a suggestion to mitigate that a bit. There also seems to be a bit of mix-up with referencing pipeline by ID vs. name, only with variable naming... all the actual code references are by ID but we should keep that crystal clear to avoid confusion in the future.

The logic looks super solid otherwise!

briancain committed:
typo: visisted => visited

briancain committed:
Fixup: Doc String PR Suggestions

Co-authored-by: Shirley Xiaolin Xu <34314221+xiaolin-ninja@users.noreply.github.com>

briancain committed:
pr feedback: Doc comment updates, var param renaming

briancain committed:
config: Add unit tests for validating pipeline steps

briancain committed:
config: Add unit test for bad syntax with both use and pipeline stanza

briancain committed:
server: Use structs to track node id and pipeline step refs

This simplifies the node id tracking in that we no longer have to loop
over the node id map to extract the node ID given a pipeline step ref.
Now instead, we can look it up as a key within a map instead of looping.

briancain committed:
Merge pull request #3751 from hashicorp/feat/cli/auto-sync-pipelines

CLI: Auto ConfigSync Pipelines

xiaolin-ninja commented on pkg/server/singleprocess/service_pipeline.go:
```diff
@@ -180,12 +180,32 @@ func (s *Service) RunPipeline(
 	req *pb.RunPipelineRequest,
 	visitedPipelines map[string]interface{},
-	nodeIdMap map[string]*pb.Ref_PipelineStep,
+	nodeStepRef *nodeToStepRef,
```
YAS BRIAN.

xiaolin-ninja reviewed (approved):
LGTM!

briancain committed:
pr feedback: Simplify stepGraph initialization
